greenpithumb/sleep_windows.py

```diff
@@ -0,0 +1,84 @@
+"""Parses sleep window strings."""
+
+import datetime
+import re
+
+
+class Error(Exception):
+    pass
+
+
+class InvalidWindowFormatError(Error):
+    pass
+
+
+class HypersomniaError(Error):
+    pass
+
+
+def parse(sleep_windows):
+    """Converts sleep windows from strings to datetime objects.
+
+    Args:
+        sleep_windows: A list of sleep window strings.Each sleep window should
+            be in 24-hour HH:MM-HH:MM format.
+
+    Returns:
+        A list of 2-tuples, each representing a sleep window. Tuple items
+        are datetime.time objects.
+    """
+    awake_minutes = [1] * 60 * 24
+    sleep_minutes = set()
+    sleep_windows_datetime = []
+    for sleep_window in sleep_windows:
+        try:
+            window_match = re.match(r'(\d\d):(\d\d)-(\d\d):(\d\d)',
+                                    sleep_window)
+        except TypeError:
+            raise InvalidWindowFormatError(
+                'Time windows must be in string format.')
+        if not window_match:
+            raise InvalidWindowFormatError(
+                'Sleep windows must be in 24-hour HH:MM-HH:MM format')
+        try:
+            sleep_time = datetime.time(
+                int(window_match.group(1)), int(window_match.group(2)))
+            wake_time = datetime.time(
+                int(window_match.group(3)), int(window_match.group(4)))
+        except (ValueError, TypeError):
+            raise InvalidWindowFormatError(
+                'Sleep windows must be in 24-hour HH:MM-HH:MM format')
+        if sleep_time == wake_time:
+            raise HypersomniaError(
+                'Sleep windows cannot cover the entire 24-hour range.')
+        sleep_windows_datetime.append((sleep_time, wake_time))
+        sleep_minutes.update(_sleep_minutes(sleep_time, wake_time))
+    for minute in sleep_minutes:
+        awake_minutes[minute] = 0
+    if not sum(awake_minutes):
+        raise HypersomniaError(
+            'Sleep windows cannot cover the entire 24-hour range.')
+
+    return sleep_windows_datetime
+
+
+def _sleep_minutes(sleep_time, wake_time):
```

Comment:
This feels like a pretty complicated way of making sure they don't specify 100% sleep time. I feel like if we have to do something complicated, that complicated thing might as well be coalescing overlapping/adjacent windows. I predict the coalesce logic will be simpler than building a set of every minute. Once we have coalesced windows, the check for hypersomnia is just checking `sleep_time == wake_time`.

The other advantage of coalescing is that it's independent of time granularity. If we at some point decide to allow seconds-level granularity, we have to change a lot of code in the current implementation, whereas with coalescing, it should just work no matter the time precision.

Alternatively, we could just not check for hypersomnia and so it's possible for the user to configure a pump that always sleeps.